File: factorial1.py
# factorial1.py

# Author: David Galilei Natale
# Last edited: May 2019

# Enter the Python interpreter and type the following commands to run:
# >>> import factorial1
# >>> factorial1.fact1(enter value)

# On my computer powered with an Intel 3.6 GHz CPU,
# 1,100,000,000! was calculated and printed in 16 days.
# This number has 9,467,808,029 digits,
# starts with 3164148161 and ends with 0000000000.

# I used Python 3.6.

import datetime
import math
import marshal

def fact1(x):
	t1 = datetime.datetime.now()

	y = math.factorial(x)

	outFile = open('NthFactorial', 'wb')
	marshal.dump(y, outFile)
	outFile.close()

	a = open('NthFactorial', 'rb')
	b = pickle.load(a)

	print ('\n\n', x ,'! has', math.trunc(math.log10(b)+1), 'digits.')

	print ('\n\nThe first 10 digits are: ', b // 10**(int(math.log(b,10)) - 10 + 1))

	print ('\n\nThe last 10 digits are: ', b//10**9%10,b//10**8%10,b//10**7%10,b//10**6%10,b//10**5%10,b//10**4%10,b//10**3%10,b//10**2%10, b//10**1%10,b//10**0%10)
		
	t2 = datetime.datetime.now()

	print (t2 - t1)

File: factorial1.c

/*                      factorial1.c

   Author: Nyna of PHP-Games (I don't know her last name.)

   Last edited: January 19, 2009

   Compiled on GNU C running on Fedora 9.

   Requires the GNU MP arbitrary precision library. I used v. 4.2.2.

   To run, on the command line type: 

   gcc factorial1.c -lgmp [Enter]
   ./a.out   

   David Natale computed 260000000!. This number has 2,074,976,510
   digits, starts out with 6083364006 and ends with 0000000000.     */
                                                                   

#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/resource.h>
#include <gmp.h>
#include <time.h>

int main( void )
{
	mpz_t z;
	FILE *outFile;
  	time_t  t0, t1; /* time_t is defined on <bits/types.h> as long */
	clock_t c0, c1; /* clock_t is defined on <bits/types.h> as long */


	t0 = time(NULL);
	c0 = clock();
	printf ("\tbegin (wall):            %ld seconds\n", (long) t0);
	printf ("\tbegin (CPU):             %ld cycles\n", (long) c0);

	mpz_init(z);
	mpz_fac_ui(z, 5);

	outFile = fopen("factorialAnswer", "w");
	gmp_fprintf(outFile, "%Zd", z);
	fclose(outFile);
	mpz_clear(z);

	t1 = time(NULL);
	c1 = clock();
	printf ("\n\tend (wall):              %ld seconds\n", (long) t1);
	printf ("\tend (CPU);               %ld cycles\n", (long) c1);
	printf ("\n\telapsed wall clock time: %ld seconds\n", (long) (t1 - t0));
	printf ("\telapsed CPU time:        %f seconds\n", (float) (c1 - c0)/CLOCKS_PER_SEC);

	return 0;
}



